hbase-client/src/main/java/org/apache/hadoop/hbase/client/CheckAndMutate.java

```diff
@@ -0,0 +1,384 @@
+     * @param value the expected value
+     * @return the CheckAndMutate object
+     */
+    public Builder ifMatches(byte[] family, byte[] qualifier, CompareOperator compareOp,
+      byte[] value) {
+      this.family = Preconditions.checkNotNull(family, "family is null");
+      this.qualifier = qualifier;
+      this.op = Preconditions.checkNotNull(compareOp, "compareOp is null");
+      this.value = value;
+      return this;
+    }
+
+    /**
+     * @param filter filter to check
+     * @return the CheckAndMutate object
+     */
+    public Builder ifMatches(Filter filter) {
+      this.filter = Preconditions.checkNotNull(filter, "filter is null");
+      return this;
+    }
+
+    /**
+     * @param timeRange time range to check
+     * @return the CheckAndMutate object
+     */
+    public Builder timeRange(TimeRange timeRange) {
+      this.timeRange = timeRange;
+      return this;
+    }
+
+    private void preCheck(Row action) {
+      Preconditions.checkNotNull(action, "action (Put/Delete/RowMutations) is null");
+      if (!Bytes.equals(row, action.getRow())) {
+        throw new IllegalArgumentException("The row of the action (Put/Delete/RowMutations) <" +
+          Bytes.toStringBinary(action.getRow()) + "> doesn't match the original one <" +
+          Bytes.toStringBinary(this.row) + ">");
+      }
+      Preconditions.checkState(op != null || filter != null, "condition is null. You need to"
+        + " specify the condition by calling ifNotExists/ifEquals/ifMatches before building a"
+        + " CheckAndMutate object");
+    }
+
+    /**
+     * @param put data to put if check succeeds
+     * @return a CheckAndMutate object
+     */
+    public CheckAndMutate build(Put put) {
+      preCheck(put);
+      if (filter != null) {
+        return new CheckAndMutate(row, filter, timeRange, put);
+      } else {
+        return new CheckAndMutate(row, family, qualifier, op, value, timeRange, put);
+      }
+    }
+
+    /**
+     * @param delete data to delete if check succeeds
+     * @return a CheckAndMutate object
+     */
+    public CheckAndMutate build(Delete delete) {
+      preCheck(delete);
+      if (filter != null) {
+        return new CheckAndMutate(row, filter, timeRange, delete);
+      } else {
+        return new CheckAndMutate(row, family, qualifier, op, value, timeRange, delete);
+      }
+    }
+
+    /**
+     * @param mutation mutations to perform if check succeeds
+     * @return a CheckAndMutate object
+     */
+    public CheckAndMutate build(RowMutations mutation) {
+      preCheck(mutation);
+      if (filter != null) {
+        return new CheckAndMutate(row, filter, timeRange, mutation);
+      } else {
+        return new CheckAndMutate(row, family, qualifier, op, value, timeRange, mutation);
+      }
+    }
+  }
+
+  /**
+   * returns a builder object to build a CheckAndMutate object
+   *
+   * @param row row
+   * @return a builder object
+   */
+  public static Builder builder(byte[] row) {
```

Comment:
nit: `newBuilder(byte[] row)`? It would match the protobuf-java API, but i'm not sure if we have a convention in HBase itself?